source/particle/integrator/rk_2.cc

```diff
@@ -41,22 +40,32 @@ namespace aspect
                                      const std::vector<Tensor<1,dim> > &velocities,
                                      const double dt)
       {
-        typename std::multimap<types::LevelInd, Particle<dim> >::iterator it = begin_particle;
-        typename std::vector<Tensor<1,dim> >::const_iterator old_vel = old_velocities.begin();
-        typename std::vector<Tensor<1,dim> >::const_iterator vel = velocities.begin();
+        Assert(std::distance(begin_particle, end_particle) == old_velocities.size(),
+               ExcMessage("The particle integrator expects the old velocity vector to be of equal size"
+                          "to the number of particles to advect. For some unknown reason they are different, "
+                          "most likely something went wrong in the calling function."));
+
+        Assert(std::distance(begin_particle, end_particle) == velocities.size(),
+               ExcMessage("The particle integrator expects the velocity vector to be of equal size"
```

Comment:
It is debug mode, I do not care ;-). 
But we do not need to make it slower than necessary, I changed it.